pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java

```diff
@@ -175,4 +183,43 @@ public String getCrcMetadataForTable(
       }
     }
   }
+
+  // TODO Add access control similar to PinotSegmentUploadDownloadRestletResource for segment download.
+  @GET
+  @Produces(MediaType.APPLICATION_OCTET_STREAM)
+  @Path("/segments/{tableNameWithType}/{segmentName}")
+  @ApiOperation(value = "Download a segment", notes = "Download a segment in zipped tar format")
+  public Response downloadSegment(
+      @ApiParam(value = "Name of the table with type REALTIME OR OFFLINE", required = true, example = "myTable_OFFLINE") @PathParam("tableNameWithType") String tableNameWithType,
+      @ApiParam(value = "Name of the segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
+      @Context HttpHeaders httpHeaders)
+      throws Exception {
+    LOGGER.info("Received a request to download segment {} for table {}", segmentName, tableNameWithType);
+    TableDataManager tableDataManager = checkGetTableDataManager(tableNameWithType);
+    SegmentDataManager segmentDataManager = tableDataManager.acquireSegment(segmentName);
+    if (segmentDataManager == null) {
+      throw new WebApplicationException(
+          String.format("Table %s segment %s does not exist", tableNameWithType, segmentName),
+          Response.Status.NOT_FOUND);
+    }
+    try {
+      String tableDir = tableDataManager.getTableDataDir();
+      // TODO Limit the number of concurrent downloads of segments because compression is an expensive operation.
+      String tarFilePath = TarGzCompressionUtils.createTarGzOfDirectory(tableDir + File.separator + segmentName);
```

Comment:
You should put all `.tar.gz` file in some temporary directory, and clean them up whenever server restarted. Otherwise, if the server shut down when downloading the segment, the segment will remain there forever

Reply:
How about adding a tarFile.deleteOnExit(); here so that the file will be deleted when the vm exits?

Reply:
You can use the temp dir that is configured as segmentTarDir for the server

Reply:
You have two options here:
(1) Create a temporary tar file (unique name), and delete it soon after you send the segment out, in the finally block. In addition you can also mark it as deleteOnExit, in case things happen during tar/serving. In this case, use segmentTarDir config to drop the file for (one) serve to another host.
(2) Create a semi-permanent tar file (non-unique name) so that it can be re-used when another request comes in for the same segment. In this case, you need to make sure that the file is not over-written when two requests come in for serving the same segment. In this case, I think the permanent tar file should be created in tableDataDir like you have done. But you do need to synchronize the creation of the file, and check if the file is already created by another thread, no?

Which technique do you want to go with?

Reply:
I would go with (1) for now. (2) goes to the same line as the performance optimization we put as TODO here. The deleteOnExit has been added already. Using segmentTarDir config seems to be an extra protection against polluting server disk.

Reply:
The segmentTarDir may be configured to be the high performance disk (e.g. ssd)  as opposed to the disk for log messages (boot/hdd). So, we should use that for sure. Please go ahead and change it to use a unique name, and add a note that this may deteriorate performance if more than one replica asks to download segments.

Reply:
Done using segmentTarDir to store the temporary file. TODO added.